src/rust/engine/client/src/build_root.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use log::debug;
+use std::env;
+use std::ops::Deref;
+use std::path::PathBuf;
+
+#[derive(Debug)]
+pub struct BuildRoot(PathBuf);
+
+impl BuildRoot {
+  const SENTINEL_FILES: [&'static str; 3] = ["pants", "BUILDROOT", "BUILD_ROOT"];
+
+  pub fn find() -> Result<BuildRoot, String> {
+    let cwd = env::current_dir().map_err(|e| format!("Failed to determine $CWD: {}", e))?;
+    let mut build_root = cwd.clone();
+    loop {
+      for sentinel in &Self::SENTINEL_FILES {
+        let sentinel_path = build_root.join(sentinel);
+        if !sentinel_path.exists() {
+          continue;
+        }
+        let sentinel_path_metadata = sentinel_path.metadata().map_err(|e| {
+          format!(
+            "\
+            Failed to read metadata for {path} to determine if is a build root sentinel file: {err}\
+            ",
+            path = sentinel_path.display(),
+            err = e
+          )
+        })?;
+        if sentinel_path_metadata.is_file() {
+          let root = BuildRoot(build_root);
+          debug!("Found {:?} starting search from {}.", root, cwd.display());
+          return Ok(root);
+        }
+      }
+
+      build_root = build_root
+        .parent()
+        .ok_or(format!(
+          "\
+          No build root detected for the current directory of {cwd}. Pants detects the build root \
+          by looking for at least one file from {sentinel_files} in the cwd and its ancestors. If \
+          you have none of these files, you can create an empty file in your build root.\
+          ",
+          cwd = cwd.display(),
+          sentinel_files = Self::SENTINEL_FILES.join(", ")
+        ))?
+        .into();
+    }
+  }
+}
+
+impl Deref for BuildRoot {
+  type Target = PathBuf;
+
+  fn deref(&self) -> &PathBuf {
+    &self.0
+  }
+}
+
+#[cfg(test)]
+mod test {
+  use crate::build_root::BuildRoot;
+  use lazy_static::lazy_static;
+  use std::ops::Deref;
```

Comment:
Hrm, this I'm not happy about. I like living in a world of rules in which case I learn them. Or a world of auto-formatters, in which case I let them do their thing. Here `cargo fmt` does only most things automatically but not this. I really don't like the idea of straddling worlds. Do you know of a tool to augment `cargo fmt` so we don't need to learn some rules and pay 1/2 attention?

Reply:
Fixed, but not happy about it. This will drift!

Reply:
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#imports_layout

We should set `imports_layout = StdExternalCrate` in the rustfmt config.

Reply:
FWIW, I'm now maintaining 3 Rust projects and this still is not in stable clippy. I just dictate those projects should be developed using `cargo +nightly fmt --all`. CI does same in check mode. This is the only +nightly; so that's a wart, but it seems totally worth it to me / gets baked into my shell history when it matters - when I'm actively hacking on those projects.